### DISCUSSION TITLE
How to provision extension configuration policies for VM?

### DISCUSSION THREAD

**Discussion by GlassGruber** [violentmonkey/violentmonkey#1482] [Q&A]
Does VM support extension policies via json config files?
What APIs are available for admin provided config files?

Suppose a bunch of machines all with Firefox browser (let'say Firefox, but could be Chrome also) all with same VM configurations and a some user scripts. How can that be done?

According to documentation linked below, an example of possible configuration file would look something like this, but what API should be called for example to provide a custom or specific usersript from some server?

```js
{
  "name": "{aecec67f-0d10-4fa7-b7c7-609a2db280cf}",
  "description": "Admin provided settings for ViolentMonkey extension",
  "type": "storage",
  "data":
  {
    //  example API 
    "userScripts": [
        "full user script json enconded string here",
        "http://remote.domain.tld/some-path/script.js"
    ]
  }
}
```

Some documentation

- https://extensionworkshop.com/documentation/enterprise/adding-policy-support-to-your-extension/

- https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/storage/managed#provisioning_managed_storage

- https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/Native_manifests#managed_storage_manifests

How uBlock origin implement this 

- https://github.com/gorhill/uBlock/wiki/Deploying-uBlock-Origin
- https://github.com/gorhill/uBlock/wiki/Deploying-uBlock-Origin:-configuration

Thank you for any suggestion!

**Comment by tophf** [comment]
It's not implemented because no one needed it and normally we don't implement niche features, especially those that aren't present in another popular userscript engine like Tampermonkey.

**Reply by GlassGruber** [reply]
Thank you for your answer.

I see, do you have any suggestions on how to deploy same configurations for VM on multiple machines without manually configure each one?

**Reply by tophf** [reply]
Since the browsers don't have a policy/method to provide the entire contents of an extension's storage when installing the extension, you'll have to write such a utility yourself or find an existing one. For example, in Chrome it's possible to write an extension that would 1) open the URL of Violentmonkey's dashboard, 2) attach to it via `chrome.debugger` API, 3) inject code that sets the storage, 4) restart Violentmonkey.

P.S. Normally enterprises shouldn't allow userscript extensions such as VM/TM as they allow bypassing security restrictions.

**Reply by GlassGruber** [reply]
> Since the browsers don't have a policy/method to provide the entire contents of an extension's storage when installing the extension, you'll have to write such a utility yourself or find an existing one. 

I would say that actually the policy framework I linked in the documents allow this type of actions, I you look at uBlock, it allows custom filters and stuff like that.

> For example, in Chrome it's possible to write an extension that would 1) open the URL of Violentmonkey's dashboard, 2) attach to it via chrome.debugger API, 3) inject code that sets the storage, 4) restart Violentmonkey.

Interesting, the debugger attach approach looks hacky but I like this type of workarounds. But considering the effort and what I said above I guess it would be wiser to extend current VM API for a policies.json setup. I'd like to do that, maybe if I can find some time I could give it a try and send a PR! Unfortunately I can already tell chances are not high but who knows.

> P.S. Normally enterprises shouldn't allow userscript extensions such as VM/TM as they allow bypassing security restrictions

Absolutely, this is a particular setup, it's more a dedicated team scenario, not really numerous. 
The policies.json allow multiple machines to be in sync with a bunch of configs and extensions, and it's quite convenient. VM is one of the extensions handled with this setup.

Thank you for your time!

**Reply by tophf** [reply]
I meant a universal method of supplying storage that doesn't require implementing it in every extension one might want to pre-configure.

**Reply by GlassGruber** [reply]
Oh sure, yea that's a bummer.
